62.CAD/Polyline.h

```diff
@@ -255,10 +281,221 @@ class CPolyline
-	// TODO[Przemek]: instead of float64_t2 for linePoints, store LinePointInfo
-	std::vector<float64_t2> m_linePoints;
+	std::vector<LinePointInfo> m_linePoints;
 	std::vector<QuadraticBezierInfo> m_quadBeziers;
+
+private:
+	class PolylineConnectorBuilder
+	{
+	public:
+		void addLineNormal(const float64_t2& line, float lineLen, const float64_t2& worldSpaceCircleCenter, float phaseShift)
+		{
+			PolylineConnectorNormalHelperInfo connectorNormalInfo;
+			connectorNormalInfo.worldSpaceCircleCenter = float32_t2(worldSpaceCircleCenter);
+			connectorNormalInfo.normal = float32_t2(-line.y, line.x) / static_cast<float>(lineLen);
+			connectorNormalInfo.phaseShift = phaseShift;
+			connectorNormalInfo.type = ObjectType::LINE;
+
+			connectorNormalInfos.push_back(connectorNormalInfo);
+		}
+
+		void addBezierNormals(const QuadraticBezierInfo& quadBezierInfo, float phaseShift)
+		{
+			// TODO: we already calculate quadratic form of each bezier (except of the last one), maybe store this info in an array and use it later to calculate normals?
+			const float64_t2 bezierDerivativeValueAtP0 = 2.0 * (quadBezierInfo.p[1] - quadBezierInfo.p[0]);
+			const float32_t2 tangentAtP0 = glm::normalize(bezierDerivativeValueAtP0);
+			//const float_t2 A = P0 - 2.0 * P1 + P2;
+			const float64_t2 bezierDerivativeValueAtP2 = 2.0 * (quadBezierInfo.p[0] - 2.0 * quadBezierInfo.p[1] + quadBezierInfo.p[2]) + 2.0 * (quadBezierInfo.p[1] - quadBezierInfo.p[0]);
+			const float32_t2 tangentAtP2 = glm::normalize(bezierDerivativeValueAtP2);
+
+			PolylineConnectorNormalHelperInfo connectorNormalInfoAtP0{};
+			connectorNormalInfoAtP0.worldSpaceCircleCenter = quadBezierInfo.p[0];
+			connectorNormalInfoAtP0.normal = float32_t2(-tangentAtP0.y, tangentAtP0.x);
+			connectorNormalInfoAtP0.phaseShift = phaseShift;
+			connectorNormalInfoAtP0.type = ObjectType::QUAD_BEZIER;
+
+			PolylineConnectorNormalHelperInfo connectorNormalInfoAtP2{};
+			connectorNormalInfoAtP2.worldSpaceCircleCenter = quadBezierInfo.p[2];
+			connectorNormalInfoAtP2.normal = float32_t2(-tangentAtP2.y, tangentAtP2.x);
+			connectorNormalInfoAtP2.phaseShift = phaseShift;
+			connectorNormalInfoAtP2.type = ObjectType::QUAD_BEZIER;
+
+			connectorNormalInfos.push_back(connectorNormalInfoAtP0);
+			connectorNormalInfos.push_back(connectorNormalInfoAtP2);
+		}
+
+		std::vector<PolylineConnector> buildConnectors()
+		{
+			std::vector<PolylineConnector> connectors;
+
+			if (connectorNormalInfos.size() < 2u)
+				return {};
+
+			uint32_t i = getConnectorNormalInfoCountOfLineType(connectorNormalInfos[1].type);
+			while (i < connectorNormalInfos.size())
+			{
+				const auto& prevLine = connectorNormalInfos[i - 1];
+				const auto& nextLine = connectorNormalInfos[i];
+
+				const float32_t2 prevLineNormal = connectorNormalInfos[i - 1].normal;
+				const float32_t2 nextLineNormal = connectorNormalInfos[i].normal;
+
+				const float crossProductZ = nextLineNormal.x * prevLineNormal.y - prevLineNormal.x * nextLineNormal.y;
```

Comment:
let's add a `cross` function in hlsl cpp compat functions that takes in `float_t2`

Reply:
do we call it smth like `calcCrossZFrom2dVec`? function called cross that takes 2d vectors and outputs a scalar may be confusing

Reply:
I usually name this `cross2D`, another thing you could do is forget about this and use the good ol' `cross(float3(v1.x,v1.y,0.0), .....)).z` but I don't suggest it cause it has redundant caluclation and we use this function a lot in 2D CAD

Refer back to this discussion here: https://discord.com/channels/593902898015109131/1116649081612292207/1166281494197706772

Reply:
will just call it cross2D